File: GILua/src/main/java/org/anime_game_servers/gi_lua/models/scene/group/SceneGadget.java
package org.anime_game_servers.gi_lua.models.scene.group;

import lombok.Getter;
import lombok.ToString;
import org.anime_game_servers.core.base.annotations.lua.LuaNames;
import org.anime_game_servers.gi_lua.models.constants.EntityType;

import java.util.List;
import java.util.Map;

@SuppressWarnings("FieldMayBeFinal")
@ToString
@Getter
public class SceneGadget extends SceneObject {
    @LuaNames("gadget_id")
    private int gadgetId;
    private int state;
    @LuaNames("point_type")
    private int pointType;
    @LuaNames("boss_chest")
    private SceneBossChest bossChest;
    @LuaNames("drop_id")
    private int dropId;
    @LuaNames("chest_drop_id")
    private int chestDropId;
    @LuaNames("interact_id")
    private int interactId;
    @LuaNames("isOneoff")
    private boolean isOneOff = false;
    @LuaNames("draft_id")
    private int draftId;
    @LuaNames("route_id")
    private int routeId;
    @LuaNames("start_route")
    private boolean startRoute = true;
    @LuaNames("is_use_point_array")
    private boolean isUsePointArray  = false;
    private boolean persistent = false;
    @LuaNames("showcutscene")
    private boolean showCutscene = false;
    private int owner; // cfg id
    private boolean autopick = false;
    private Explore explore;
    private List<Integer> arguments;
    @LuaNames("oneoff_reset_version")
    private int oneOffResetVersion;
    @LuaNames("is_guest_can_operate")
    private boolean isGuestCanOperate = false;
    @LuaNames("is_blossom_chest")
    private boolean isBlossomChest = false;
    @LuaNames("is_enable_interact")
    private boolean isEnableInteract = true;
    @LuaNames("talk_state")
    private int talkState;
    @LuaNames("fishing_id")
    private int fishingId;
    @LuaNames("fishing_areas")
    private List<Integer> fishingAreas;
    @LuaNames("crucible_config")
    private CrucibleConfig crucibleConfig;
    @LuaNames("offering_config")
    private OfferingConfig offeringConfig;
    @LuaNames("worktop_config")
    private WorktopConfig worktopConfig;
    @LuaNames("server_global_value_config")
    private Map<String, Float> serverGlobalValueConfig;

    @Override
    public EntityType getType() {
        return EntityType.GADGET;
    }
}

File: GILua/src/main/java/org/anime_game_servers/gi_lua/models/scene/group/SceneGadget.kt
package org.anime_game_servers.gi_lua.models.scene.group

import org.anime_game_servers.core.base.annotations.lua.LuaNames
import org.anime_game_servers.gi_lua.models.constants.EntityType
import org.anime_game_servers.gi_lua.models.constants.VisionLevelType

data class SceneGadget(
    @field:LuaNames("gadget_id")
    val gadgetId: Int = 0,
    val state: Int = 0,

    @field:LuaNames("point_type")
    val pointType: Int = 0,

    @field:LuaNames("boss_chest")
    val bossChest: SceneBossChest? = null,


    @field:LuaNames("chest_drop_id")
    val chestDropId: Int = 0,

    @field:LuaNames("interact_id")
    val interactId: Int = 0,

    @field:LuaNames("draft_id")
    val draftId: Int = 0,

    @field:LuaNames("route_id")
    val routeId: Int = 0,

    @field:LuaNames("start_route")
    val startRoute: Boolean = true,

    @field:LuaNames("is_use_point_array")
    val isUsePointArray: Boolean = false,
    val persistent: Boolean = false,

    @field:LuaNames("showcutscene")
    val showCutscene: Boolean = false,
    val owner: Int = 0, // cfg id
    val autopick: Boolean = false,
    val explore: Explore? = null,
    val arguments: MutableList<Int?>? = null,

    @field:LuaNames("is_guest_can_operate")
    val isGuestCanOperate: Boolean = false,

    @field:LuaNames("is_blossom_chest")
    val isBlossomChest: Boolean = false,

    @field:LuaNames("is_enable_interact")
    val isEnableInteract: Boolean = true,

    @field:LuaNames("talk_state")
    val talkState: Int = 0,

    @field:LuaNames("fishing_id")
    val fishingId: Int = 0,

    @field:LuaNames("fishing_areas")
    val fishingAreas: MutableList<Int?>? = null,

    @field:LuaNames("crucible_config")
    val crucibleConfig: CrucibleConfig? = null,

    @field:LuaNames("offering_config")
    val offeringConfig: OfferingConfig? = null,

    @field:LuaNames("worktop_config")
    val worktopConfig: WorktopConfig? = null,


    override val level: Int = 0,
    @field:LuaNames("vision_level")
    override val visionLevel: VisionLevelType = VisionLevelType.VISION_LEVEL_NORMAL,
    @field:LuaNames("mark_flag")
    override val markFlag: Int = 0,
    @field:LuaNames("isOneoff")
    override val isOneOff: Boolean = false,
    @field:LuaNames("oneoff_reset_version")
    override val oneOffResetVersion: Int = 0,
    @field:LuaNames("drop_tag")
    override val dropTag: String? = null,
    @field:LuaNames("drop_id")
    override val dropId: Int = 0,

    @field:LuaNames("guest_ban_drop")
    override val guestBanDrop: Int = 0,
    @field:LuaNames("server_global_value_config")
    override val serverGlobalValueConfig: MutableMap<String, Float>? = null,
) : SceneObject(), SceneCreature // todo add parent fields to constructor
{
    override val type: EntityType = GADGET
}
